sdk/servicebus/service-bus/src/track2/serviceBusReceiverClient.ts

```diff
@@ -0,0 +1,568 @@
+  >;
+
+  /**
+   * Creates a client for an Azure Service Bus queue.
+   *
+   * @param subscriptionAuth Data needed to connect to a subscription.
+   * @param receiveMode The receive mode to use (defaults to PeekLock)
+   * @param options Options for the client itself.
+   */
+  new (
+    subscriptionAuth: SubscriptionAuth,
+    receiveMode: "receiveAndDelete",
+    session: Session
+  ): ClientTypeT<"receiveAndDelete", "subscription", "sessions">;
+}
+
+/**
+ * Implementation class for receivers.
+ * @internal
+ * @ignore
+ */
+export class ReceiverClientImplementation {
+  constructor(
+    auth1: QueueAuth | SubscriptionAuth,
+    receiveMode2: "peekLock" | "receiveAndDelete",
+    sessionOrOptions3?: Session | ServiceBusClientOptions,
+    options4?: ServiceBusClientOptions
+  ) {
+    let options: ServiceBusClientOptions;
+    let session: Session | undefined;
+
+    if (sessionOrOptions3 != null && isSession(sessionOrOptions3)) {
+      session = sessionOrOptions3;
+      options = options4 || {};
+    } else {
+      options = sessionOrOptions3 || {};
+    }
+
+    const { context, entityPath } = createConnectionContext(auth1, options);
+    this._context = context;
+    this._receiveMode = convertToInternalReceiveMode(receiveMode2);
+
+    const clientEntityContext = ClientEntityContext.create(
+      entityPath,
+      ClientType.ServiceBusReceiverClient,
+      context,
+      `${entityPath}/${generate_uuid()}`
+    );
+
+    // TODO: use the session connections object to "cache" the client entity context
+    if (session != null) {
+      const receiver = new InternalSessionReceiver(clientEntityContext, this._receiveMode, {
+        sessionId: session.id
+      });
+      this._sessionEnabled = true;
+      this._receiver = receiver;
+
+      this.diagnostics = {
+        async peek(maxMessageCount?: number): Promise<ReceivedMessage[]> {
+          return (await receiver.peek(maxMessageCount)).map((m) => m as ReceivedMessage);
+        },
+        async peekBySequenceNumber(
+          fromSequenceNumber: Long,
+          maxMessageCount?: number
+        ): Promise<ReceivedMessage[]> {
+          return (await receiver.peekBySequenceNumber(fromSequenceNumber, maxMessageCount)).map(
+            (m) => m as ReceivedMessage
+          );
+        }
+      };
+    } else {
+      const receiver = new InternalReceiver(clientEntityContext, this._receiveMode);
+      this._sessionEnabled = false;
+      this._receiver = receiver;
+
+      this.diagnostics = {
+        async peek(maxMessageCount?: number): Promise<ReceivedMessage[]> {
+          return (await receiver.peek(entityPath, maxMessageCount)).map(
+            (m) => m as ReceivedMessage
+          );
+        },
+        async peekBySequenceNumber(
+          fromSequenceNumber: Long,
+          maxMessageCount?: number
+        ): Promise<ReceivedMessage[]> {
+          return (
+            await receiver.peekBySequenceNumber(entityPath, fromSequenceNumber, maxMessageCount)
```

Comment:
My own curiosity, does entityPath have the session identifier and that's why it's needed to be passed in here? (Does the receiver already have the entityPath?)

Reply:
I'd like to look at that more in a later PR. 

It looks like it's cosmetic but I'm curious why it doesn't grab it from it's own internal context at that point (which should _also_ have the same entity path in it).